core/src/main/java/com/linecorp/armeria/common/HttpRequestBuilder.java

```diff
@@ -0,0 +1,417 @@
+     * }</pre>
+     */
+    public HttpRequestBuilder pathParams(Map<String, Object> pathParams) {
+        requireNonNull(pathParams, "pathParams");
+        this.pathParams.putAll(pathParams);
+        return this;
+    }
+
+    /**
+     * Sets a query param for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/endpoint")
+     *            .queryParam("foo", "bar")
+     *            .build(); // GET `/endpoint?foo=bar`
+     * }</pre>
+     */
+    public HttpRequestBuilder queryParam(String name, Object value) {
+        requireNonNull(name, "name");
+        requireNonNull(value, "value");
+        queryParamsBuilder.setObject(name, value);
+        return this;
+    }
+
+    /**
+     * Sets multiple query params for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/endpoint")
+     *            .queryParams(QueryParams.of("from", "foo", "limit", 10))
+     *            .build(); // GET `/endpoint?from=foo&limit=10`
+     * }</pre>
+     *
+     * @see QueryParams
+     */
+    public HttpRequestBuilder queryParams(QueryParams queryParams) {
+        requireNonNull(queryParams, "queryParams");
+        for (String name : queryParams.names()) {
+            queryParamsBuilder.set(name, requireNonNull(queryParams.get(name), "query"));
+        }
+        return this;
+    }
+
+    /**
+     * Sets a cookie for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/")
+     *            .cookie(Cookie.of("cookie", "foo"))
+     *            .build();
+     * }</pre>
+     *
+     * @see Cookie
+     */
+    public HttpRequestBuilder cookie(Cookie cookie) {
+        requireNonNull(cookie, "cookie");
+        cookies.add(cookie);
+        return this;
+    }
+
+    /**
+     * Sets multiple cookies for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/")
+     *            .cookies(Cookies.of(Cookie.of("cookie1", "foo"),
+     *                                Cookie.of("cookie2", "bar")))
+     *            .build();
+     * }</pre>
+     *
+     * @see Cookies
+     */
+    public HttpRequestBuilder cookies(Cookies cookies) {
+        requireNonNull(cookies, "cookies");
+        this.cookies.addAll(cookies);
+        return this;
+    }
+
+    /**
+     * Creates a new {@link HttpRequest}.
+     */
+    public HttpRequest build() {
+        final RequestHeaders requestHeaders = requestHeaders();
+        if (content == null || content.length() == 0) {
+            if (content != null) {
+                content.close();
+            }
+            if (publisher != null) {
+                if (publisher instanceof HttpRequest) {
+                    return ((HttpRequest) publisher).withHeaders(requestHeaders);
+                }
+                return new PublisherBasedHttpRequest(requestHeaders, publisher);
```

Comment:
If `httpTrailersBuilder` is not empty, we need to add the trailers as the last part of `publisher`. This is complex a bit.
I was working on making a StreamMessage that append a message to an existing StreamMessage if it was completed successfully.

https://github.com/ikhoon/armeria/blob/multipart/core/src/main/java/com/linecorp/armeria/common/multipart/StreamMessageOnCompleteResumeWith.java

I think we can support `Publisher` later when #2894 is merged,

Reply:
Sounds good, let's push back `Publisher` now. I'll add a TODO.